### ISSUE TITLE
Composability

### ISSUE THREAD

**Issue by ianhi** [ianhi/mpl-interactions#117]
### Problem

You currently can't mix interactive functions. So 
```python
interactive_plot(...)
interactive_scatter(...)
```
is impossible. Or actually it is possible but it will result in multiple figure draws because  every update function always calls `fig.canvas.draw()` I thought about using an argument like `draw` to allow toggling this:
```python
interactive_plot(..., draw = False)
interactive_hist(..., draw = False)
interactive_scatter(..., draw = True)
```
but this kinda sucks. It makes you have to think about what the last call will be and it isn't very matplotlib-like.

The other issue that's been blocking this is how to easily share controls between different calls becuase currently they are just returned as an unstructured list

### Proposed Solution

Create a `Controls` object that allows news functions to register with it and keeps track of all the figures it affects. This object will make sure to call all of the update functions before calling `draw`. It should also be passable as a `controls = ` argument to subsequent calls to interactive functions. and then those functions can just register themselves to it. Along with this the pyplot funcitons should lean into the fact that they are pyplot functions and call `gca` just like real pyplot. So this should be possible:

```python
fig, ax = plt.subplots()
controls = interactive_plot(f_line,  param1 = (0, 10), param2 = (-5, 5), param3 = 5)
interactvie_hist(f_hist, controls = controls)
interactive_scatter(f_scat, controls = controls)
```

Furthermore, new functions should be able to add more sliders to the controller but it should not apply these controls to the earlier sliders.
So this should work:
```python
fig, ax = plt.subplots()
def f_line(param1, param2, param3):
...
def f_hist(param1, param2, param3, param4):
...
controls = interactive_plot(f_line,  param1 = (0, 10), param2 = (-5, 5), param3 = 5)
interactive_hist(f_hist, controls = controls, param4 = (-1,0))
```

**using fewer sliders**
It should also be possible for future interactive functions to use fewer parameters. I think there are two ways to accomplish this
1. using `inspect` and trying to make a good guess
2. creating a good way to specify 

approach 1
```python
fig, ax = plt.subplots()
def f_line(param1, param2, param3):
...
def f_hist(param1, param2, param3, param4):
...
controls = interactvie_hist(f_hist, param1 = (0, 10), param2 = (-5, 5), param3 = 5, param4 = (-1,0))
interactive_plot(f_line,  controls = controls)
```

approach 2

make controls return something sensible when indexed.

```python
controls = interactvie_hist(f_hist, param1 = (0, 10), param2 = (-5, 5), param3 = 5, param4 = (-1,0))
interactive_plot(f_line,  controls = controls['param1', 'param2', 'param3'])
```

I think approach 1 has the potential to be a nicer user experience, but it also seems radically more fraught with development difficulties. Perhaps I should start with approach two and implement 1 later on. 

### Additional context

This is definitely inspired by how [animatplot](https://animatplot.readthedocs.io/en/stable/index.html) does things. I think the api I'm proposing is a bit different in that animatplot saves the returns from each function and then passes them into a `amp.Animation` object to collect them all together. 

The other differences are:
- Animatplotlib seems to only support a single time axis
- amp works with arrays rather than functions
   - Different use case, not worse or better
- I think animatplotlib only works with mpl sliders
- amp has a sick play pause feature by using `FuncAnimation` i.e. they implemented a `Play` button out of matplotlib which is dope.

**Comment by ianhi** [comment]
`Controls` objects should know how to display them selves. And could even obey `ion/ioff` as to whether they display `ipywidgets`.

I think they should also be in charge of creating the sliders and and other controls

**Comment by ianhi** [comment]
working on this on the controller branch:
https://github.com/ianhi/mpl-interactions/tree/controller

This is clearly a better way to do things. v excite

**Comment by ianhi** [comment]
fixed by #119
